iothub_client/samples/iothub_ll_c2d_sample/iothub_ll_c2d_sample.c

```diff
@@ -43,13 +43,53 @@
 #include "certs.h"
 #endif // SET_TRUSTED_CERT_IN_SAMPLES
 
+// Uncomment this define to use Asynchronous ACK of Cloud-to-Device messages.
+// #define USE_C2D_ASYNC_ACK
+
+#ifdef USE_C2D_ASYNC_ACK
+#include "azure_c_shared_utility/singlylinkedlist.h"
+#endif
+
 /* Paste in the your iothub connection string  */
 static const char* connectionString = "[device connection string]";
 
 #define MESSAGE_COUNT        3
 static bool g_continueRunning = true;
 static size_t g_message_recv_count = 0;
 
+#ifdef USE_C2D_ASYNC_ACK
+static SINGLYLINKEDLIST_HANDLE g_cloudMessages;
+
+static bool ack_and_remove_message(const void* item, const void* match_context, bool* continue_processing)
+{
+    IOTHUB_DEVICE_CLIENT_LL_HANDLE device_ll_handle = (IOTHUB_DEVICE_CLIENT_LL_HANDLE)match_context;
+    IOTHUB_MESSAGE_HANDLE message = (IOTHUB_MESSAGE_HANDLE)item;
+
+    const char* messageId;
+    if ((messageId = IoTHubMessage_GetMessageId(message)) == NULL)
+    {
+        messageId = "<unavailable>";
+    }
+
+    (void)printf("Sending ACK for cloud message (Message ID: %s)\r\n", messageId);
+
+    if (IoTHubDeviceClient_LL_SendMessageDisposition(device_ll_handle, message, IOTHUBMESSAGE_ACCEPTED) != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("ERROR: IoTHubDeviceClient_LL_SendMessageDisposition..........FAILED!\r\n");
```

Comment:
delete the message here? and same comments as before about returning true/false etc

Reply:
This is a call like any other to the public API, if it fails it should print a message so the user knows something did not work.